Add hashlib-like Checksum helper class.

tseaver opened the pull request.

tseaver committed:
Add hashlib-like Checksum helper class.

dhermes commented on src/crc32c/__init__.py:
```diff
@@ -12,6 +12,8 @@
 
 # NOTE: ``__config__`` **must** be the first import because it (may)
 #       modify the search path used to locate shared libraries.
+import struct
```
@tseaver Can you move this **above** the comment about importing `__config__`?

tseaver replied:
7b9fa27

dhermes commented on src/crc32c/__init__.py:
```diff
@@ -42,3 +44,57 @@ def value(chunk):
+
+class Checksum(object):
+    """Hashlib-alike helper for CRC32C operations."""
+    __slots__ = ('_crc',)
```
Can you run `black --line-length=79 $(git ls-files -- '*.py')`?

tseaver replied:
I can live with some of its choices (`"`, meh), but reformatting the table from the RFC?  What is the comment syntax to bail out of that?

dhermes replied:
There is no bail-out for `black`. It's purpose is to avoid such arguments.

tseaver replied:
d6d73fd adds the example comments from the RFC.  6361bb5 blackens the two files.

dhermes commented on src/crc32c/__init__.py:
```diff
@@ -42,3 +44,57 @@ def value(chunk):
+    def __init__(self, initial_value=b''):
+        """Construct the helper.
+
+        Args:
```
Typically we don't put a docstring on `__init__` and put stuff like this in the class docstring

tseaver replied:
7b9fa27

dhermes commented on src/crc32c/__init__.py:
```diff
@@ -42,3 +44,57 @@ def value(chunk):
+
+        Args:
+            initial_value (Optional[bytes]): the initial chunk of data from
+            which the CRC32C checksum is computed.  Defaults to b''.
```
```suggestion
                which the CRC32C checksum is computed.  Defaults to b''.
```

tseaver replied:
7b9fa27

dhermes commented on src/crc32c/__init__.py:
```diff
@@ -42,3 +44,57 @@ def value(chunk):
+        """
+        self._crc = value(initial_value)
+
+    def __int__(self):
```
Why is this needed?

tseaver replied:
I'm using it for test assertions.  I can use the private variable instead, if you prefer.

dhermes replied:
I'd prefer, just to have a smaller code footprint here.

tseaver replied:
89a8520

dhermes commented on src/crc32c/__init__.py:
```diff
@@ -42,3 +44,57 @@ def value(chunk):
+
+        Args:
+            chunk (Optional[bytes]): a chunk of data used to extend
+            the CRC32C checksum.
```
```suggestion
                the CRC32C checksum.
```

tseaver replied:
7b9fa27

dhermes commented on src/crc32c/__init__.py:
```diff
@@ -42,3 +44,57 @@ def value(chunk):
+        return struct.pack('>L', self._crc)
+
+    def hexdigest(self):
+        """Like digest() except returned as a bytestring of double length.
```
```suggestion
        """Like :meth:`digest` except returned as a bytestring of double length.
```

tseaver replied:
7b9fa27

dhermes commented on src/crc32c/__init__.py:
```diff
@@ -42,3 +44,57 @@ def value(chunk):
+    def digest(self):
+        """Big-endian order, per RFC 4960.
+
+        Per: https://cloud.google.com/storage/docs/json_api/v1/objects#crc32c
```
This is vague. What is "Per:" in reference to?

tseaver replied:
7b9fa27

tseaver committed:
Review comments:

- Move stdlib import above comment about importing '__config__' first.
- Move __init__ arg docs to class docstring.
- Napoleon / Sphinx fixups.

tseaver committed:
Annotate CRC example data with examples from RFC 3720, B.4.

Name each case and its corresponding CRC value.

tseaver committed:
Blacken.

dhermes reviewed (approved):
LGTM

tseaver committed:
Review comment: remove 'Checksum.__int__'.
